### ISSUE TITLE
Observers

### ISSUE THREAD

**Issue by Ukilele** [jbcoe/indirect_value#75]
Hello! The Observers ([link](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p1950r1.html#xy7-class-template-indirect_value-observers-indirect_valueobservers)), currently listed in the proposal, are:
```
T& operator*();
T* operator->();
const T& operator*() const;
const T* operator->() const;
explicit operator bool() const noexcept;
```
Being inspired by [std::optional](https://en.cppreference.com/w/cpp/utility/optional) and [std::unique_ptr](https://en.cppreference.com/w/cpp/memory/unique_ptr/), I suggest to extend the Observers to contain the following:
```
T* operator->() noexcept;
const T* operator->() const noexcept;
T& operator*() & noexcept;
const T& operator*() const& noexcept;
T&& operator*() && noexcept;
const T&& operator*() const&& noexcept;

T& value() &;
const T& value() const&;
T&& value() &&;
const T&& value() const&&;

explicit operator bool() const noexcept;
bool has_value() const noexcept;

copier_type& get_copier() noexcept;
const copier_type& get_copier() const noexcept;
deleter_type& get_deleter() noexcept;
const deleter_type& get_deleter() const noexcept;
```
To summarise the differences:

1. `operator->` and `operator*` are `noexcept`
2. `operator*` is overloaded on ref-qualifiers
3. Add `value()`. Similar like [std::optional::value](https://en.cppreference.com/w/cpp/utility/optional/value), it performs a null-check and throws a `bad_indirect_value_acess` (which we would also have to add), in case that `*this` does not hold a value.
4. Add `has_value()`, which is simply a different name for spelling `bool(*this)`.
5. Add `get_copier()` and `get_deleter()`. This would require adding the member types `copier_type` and `deleter_type` to `indirect_value`. Just like this: `using copier_cype = C; using deleter_type = D;`.

**Comment by jbcoe** [comment]
These sounds like good improvements.

Are you happy to add them to the reference implementation?
